<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [332]:
import pandas as pd
import psycopg2

In [333]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [334]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [335]:
# текст запроса
# считаем общее количество вакансий
query_3_1 = f'''select COUNT(id)
                from vacancies
            '''

In [336]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [337]:
# текст запроса
# считаем общее оличество работодателей
query_3_2 = f'''select COUNT(id)
                from employers
            '''

In [338]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [339]:
# текст запроса
# считаем количество регионов
query_3_3 = f'''select COUNT(id)
                from areas
            '''

In [340]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [341]:
# текст запроса
# считаем количество сфер деятельности
query_3_4 = f'''select COUNT(id)
                from industries
            '''

In [342]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***Выводы по предварительному анализу вакансий**\
Всего вакансий - 49197,\
работодателей - 1362,\
городов, где размещены вакансии - 1362, цифра меньше, чем число городов в России(1118), следовательно, вакансии размещены и в других странах,\
сфер деятельности вакансий - 294.*

***

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [343]:
# текст запроса
# считаем количество вакансий в каждом регионе, отсортировав по убыванию
query_4_1 = f'''select a.name, COUNT(v.id)
                from areas a
                JOIN vacancies v ON v.area_id = a.id
                GROUP BY a.name
                ORDER BY COUNT(v.id) DESC
                
            '''

In [344]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


name  count
0             Москва   5333
1    Санкт-Петербург   2851
2              Минск   2112
3        Новосибирск   2006
4             Алматы   1892
..               ...    ...
764       Тарко-Сале      1
765    Новоаннинский      1
766            Бирск      1
767           Сасово      1
768             Анна      1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [345]:
# текст запроса
# считаем количество вакансий с заполненым полем ЗП
query_4_2 = f'''select COUNT(v.id)
                from vacancies v
                WHERE (v.salary_from IS NOT NULL)
                    OR (v.salary_to IS NOT NULL)
            '''

In [346]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [347]:
# текст запроса
# определяем нижнюю и верхнюю границы ЗП
query_4_3 = f'''select ROUND(AVG(v.salary_from)) as "нижняя граница", ROUND(AVG(v.salary_to)) as "верхняя граница"
                from vacancies v
            '''

In [348]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


нижняя граница  верхняя граница
0         71065.0         110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [349]:
# текст запроса
# выводим варианты сочетаний рабочего графика и типа трудоустройства, сортируем по количеству вакансий
query_4_4 = f'''select v.schedule, v.employment, COUNT(v.id)
                from vacancies v
                GROUP BY v.schedule, v.employment
                ORDER BY 3 DESC          
            '''

In [350]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [351]:
# текст запроса
# считаем количества вакансий в зависимости от опыта работы
query_4_5 = f'''select v.experience, COUNT(v.id)
                from vacancies v
                GROUP BY v.experience
                ORDER BY 2          
            '''

In [352]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

***Выводы по детальному анализу вакансий**\
Топ 5 вакансий размещены в тройке городов-миллионников России: Москва(5333), Санкт-Петербург(2851), Новосибирск(2006). Также в пятерку входят крупнейшие города своих стран: Минск(2112) - столица Беларусии и Алматы(1892), Казахстан.\
Половина вакансий размещены без указания ЗП(24073 из 49197).\
Нижняя граница ЗП - 71065р, вероятно, обусловлена высокой ЗП в крупных городах и их преобладании в общем количестве размещенных вакансий. Верхняя граница - 110537р.\
Самое популярное сочетание рабочий график-тип трудоустройства - Полный день-Полная занятость.\
Требуемый опыт работы в порядке возрастания количества вакансий: Более 6 лет(26152) — Нет опыта(14511) — От 3 до 6 лет(7197) — От 1 года до 3 лет(1337). Большой опыт работы играет немаловажную роль при выборе соискателя.*

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [353]:
# текст запроса
# выводим топ-5 работодателей по количеству вакансий
query_5_1 = f'''select e.name, COUNT(v.id)
                from employers e
                JOIN vacancies v ON v.employer_id = e.id
                GROUP BY e.name
                ORDER BY COUNT(v.id) DESC
                LIMIT 5
            '''

In [354]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [355]:
# текст запроса
# считаем кол-во работодателей, в котором нет вакансий, по регионам
query_5_2 = f'''select a.name, COUNT(DISTINCT e.id) as "количество работодателей", COUNT(DISTINCT v.id) as "количество вакансий"
                from areas a
                LEFT JOIN employers e ON a.id = e.area
                LEFT JOIN vacancies v ON a.id = v.area_id
                WHERE v.id IS NULL
                GROUP BY a.id
                ORDER BY 2 DESC, 3
            '''

In [356]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name  количество работодателей  количество вакансий
0                Россия                       410                    0
1             Казахстан                       207                    0
2    Московская область                        75                    0
3    Краснодарский край                        19                    0
4    Ростовская область                        18                    0
..                  ...                       ...                  ...
588             Скидель                         0                    0
589          Белоозерск                         0                    0
590           Орловский                         0                    0
591          Павловская                         0                    0
592             Таксимо                         0                    0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [357]:
# текст запроса
# выводим работодателей в порядке уменьшения регионов, где он публикует вакансии
query_5_3 = f'''select e.name, COUNT(DISTINCT v.area_id) as "количество регионов"
                from employers e
                JOIN vacancies v ON v.employer_id = e.id
                GROUP BY e.id
                ORDER BY 2 DESC
                LIMIT 5
            '''

In [358]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  количество регионов
0                  Яндекс                  181
1              Ростелеком                  152
2              Спецремонт                  116
3  Поляков Денис Иванович                   88
4               ООО ЕФИН                    71

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [359]:
# текст запроса
# считаем работодателей, у которых не указана сфера деятельности
query_5_4 = f'''select COUNT(DISTINCT e.id)
                from employers e
                LEFT JOIN employers_industries i ON i.employer_id = e.id
                WHERE i.industry_id IS NULL
            '''

In [360]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [361]:
# текст запроса
# выводим 3 компании, у которых указано  сферы деятельности
query_5_5 = f'''select e.name
                from employers e
                LEFT JOIN employers_industries i ON i.employer_id = e.id
                GROUP BY e.name
                HAVING COUNT(i.industry_id) = 4
                ORDER BY 1
                LIMIT 3
            '''

In [362]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name
0  101 Интернет
1      21vek.by
2          2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [363]:
# текст запроса
# считаем кол-во работодателей, чьей сферой деятельности является разработка рограммного обеспечения
query_5_6 = f'''select COUNT(DISTINCT e.id)
                from employers_industries ei
                JOIN employers e ON ei.employer_id = e.id
                JOIN industries i ON ei.industry_id = i.id
                WHERE i.name LIKE 'Разработка программного обеспечения'
            '''

In [364]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [365]:
# текст запроса
# выводим список городов-миллионников с числом вакансий, где Яндекс их размещает, предварительно обозначив список городов
cities = ('Москва', 'Санкт-Петербург', 'Новосибирск','Екатеринбург','Казань','Нижний Новгород','Красноярск','Челябинск','Самара','Уфа','Ростов-на-Дону','Краснодар','Омск','Воронеж','Пермь','Волгоград')
query_5_7 = f'''WITH temp AS(
                    select a.name, COUNT(v.id) cnt
                    from vacancies v
                    JOIN areas a ON v.area_id = a.id
                    JOIN employers e ON v.employer_id = e.id
                    WHERE (e.name LIKE 'Яндекс') AND (a.name in {cities})
                    GROUP BY a.name
                    ORDER BY 2
                    )   
                
                select * from temp
                UNION ALL
                select 'Total', SUM(cnt) from temp
            '''

In [366]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\1965596796.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name    cnt
0              Омск   21.0
1         Челябинск   22.0
2        Красноярск   23.0
3         Волгоград   24.0
4             Пермь   25.0
5            Казань   25.0
6    Ростов-на-Дону   25.0
7               Уфа   26.0
8            Самара   26.0
9         Краснодар   30.0
10          Воронеж   32.0
11      Новосибирск   35.0
12  Нижний Новгород   36.0
13     Екатеринбург   39.0
14  Санкт-Петербург   42.0
15           Москва   54.0
16            Total  485.0

***

***Выводы по анализу работодателей**\
Топ 5 работодателей по количеству размещенных вакансий - крупные, быстро развивающиеся организации: Яндекс(1933), Ростелеком(491), Тинькофф(444), СБЕР(428), Газпром нефть(331). Можно предположить, что они в большом количестве предлагают второй по популярности тип графика работы - Удаленная работа, набирающая популярность в последнее время.\
Среди регионов, в которых нет вакансий, наибольшее количество работодателей в России.\
По количеству регионов, в которых работодатели публикуют вакансии, первые 2 места в рейтинге занимают Яндекс и Ростелеком. Это можно объяснить самим масштабом данных компаний, а также возможностью удаленной работы в них (возможно).\
Работодателей, у которых не указана сфера детельности - 8419.\
Название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности - 2ГИС.\
В качестве сферы деятельности указана «Разработка программного обеспечения» у 3553 работодателей.\
"Яндекс" размещает вакансии в 16 городах-миллионниках, т.е. во всех. Всего размещено 485 вакансий в этих городах.*

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [367]:
# текст запроса
# считаем количество вакансий, имеющих отношение к данным
query_6_1 = f'''select COUNT(v.id)
                from vacancies v 
                WHERE lower(v.name) LIKE '%data%' OR lower(v.name) LIKE '%данн%'
            '''

In [368]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [369]:
# текст запроса
# считаем кол-во вакансий для начинаеющего дата-саентиста
query_6_2 = f'''select COUNT(v.id)
                from vacancies v 
                WHERE (v.name iLIKE '%data scientist%'
                    OR v.name iLIKE '%data science%'
                    OR v.name iLIKE '%исследователь данных%'
                    OR v.name iLIKE '%machine learning%'
                    OR v.name iLIKE '%машинн%обучен%'
                    OR (v.name LIKE '%ML%' AND v.name NOT ILIKE '%html%'))
                    AND
                    (v.name iLIKE '%junior%'
                    OR v.experience iLIKE 'Нет опыта'
                    OR v.employment iLIKE 'Стажировка')
            '''

In [370]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [371]:
# текст запроса
# считаем вакансии для DS, в которых в качестве ключевого навыка указан SQL или postgres
query_6_3 = f'''select COUNT(v.id)
                from vacancies v 
                WHERE (v.name iLIKE '%data scientist%'
                    OR v.name iLIKE '%data science%'
                    OR v.name iLIKE '%исследователь данных%'
                    OR v.name iLIKE '%machine learning%'
                    OR v.name iLIKE '%машинн%обучен%'
                    OR (v.name LIKE '%ML%' AND v.name NOT ILIKE '%html%'))
                    AND
                    (v.key_skills iLIKE '%SQL%'
                    OR v.key_skills iLIKE '%postgres%')
            '''

In [372]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [373]:
# текст запроса
# считаем вакансии для DS, в которых в качестве ключевого навыка указан Python
query_6_4 = f'''select COUNT(v.id)
                from vacancies v 
                WHERE (v.name iLIKE '%data scientist%'
                    OR v.name iLIKE '%data science%'
                    OR v.name iLIKE '%исследователь данных%'
                    OR v.name iLIKE '%machine learning%'
                    OR v.name iLIKE '%машинн%обучен%'
                    OR (v.name LIKE '%ML%' AND v.name NOT ILIKE '%html%'))
                    AND
                    (v.key_skills iLIKE '%Python%')
            '''

In [374]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [375]:
# текст запроса
# считаем среднее кол-во навыков, указанных в вакансиях для DS
query_6_5 = f'''select AVG(length(v.key_skills)-length(replace(v.key_skills, CHR(9), ''))+1)
                from vacancies v 
                WHERE (v.name iLIKE '%data scientist%'
                    OR v.name iLIKE '%data science%'
                    OR v.name iLIKE '%исследователь данных%'
                    OR v.name iLIKE '%machine learning%'
                    OR v.name iLIKE '%машинн%обучен%'
                    OR (v.name LIKE '%ML%' AND v.name NOT ILIKE '%html%'))
                    AND v.key_skills IS NOT NULL
            '''

In [376]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df


C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\4273863405.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


avg
0  6.406032

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [377]:
# текст запроса
# считаем среднюю ЗП в зависимости от требуемого опыта работы
query_6_6 = f'''select v.experience, round(AVG(coalesce((v.salary_from + v.salary_to)/2, v.salary_from, v.salary_to)))
                from vacancies v 
                WHERE (v.name iLIKE '%data scientist%'
                    OR v.name iLIKE '%data science%'
                    OR v.name iLIKE '%исследователь данных%'
                    OR v.name iLIKE '%machine learning%'
                    OR v.name iLIKE '%машинн%обучен%'
                    OR (v.name LIKE '%ML%' AND v.name NOT ILIKE '%html%'))
                    AND (v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL)
                GROUP BY 1     
            '''

In [378]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df


C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\3564239429.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience     round
0           Нет опыта   74643.0
1  От 1 года до 3 лет  139675.0
2       От 3 до 6 лет  243115.0

***

***Выводы по предметному анализу**\
Среди всех вакансий отношение к данным имеют 1771.\
Для начинающего дата-сайентиста(Junior) размещены 51 подходящая вакансия.\
Вакансий для всех уровней DS, в которых в качестве ключевого навыка указан SQL или postgres - 201.\
Вакансий для всех уровней DS, в которых в качестве ключевого навыка указан Pythons - 351.\
Средее количество ключевых навыков для всех уровней DS - 6.41.\
Рассматривая ЗП дата-саентиста в зависимости от опыта работы, средние значения будут следующими: Нет опыта - 74643р, От 1 года до 3 лет - 139675р, От 3 до 6 лет - 243115р.*

# Общий вывод по проекту

***Подведем итог исследования**\
Проведя дополнительные исследования ниже, ко всему выше сказанному можем добавить следующее:\
- топ 10 вакансий по количеству размещений так или иначе связаны с IT, среди них часто вречаются аналитики, программисты.\
- проверим гипотезу, что крупные, активно развивающиеся организации все чаще предлагают вакансии с удаленным типом рабочего графика. в топ 10 работодателей видим работодателей, которых мы уже встречали в пункте анализа 5.1, а именно Яндекс, Тинькофф, Ростелеком.\
- проанализировав работодателей, которые размещают достаточное количество вакансий для дата-саентистов снова видим активно развивающиеся на данный момент организации: СБЕР, ВТБ, VK, Яндекс, Мегафон, МТС.*

In [380]:
query_7_1 = f'''select name, COUNT(id)
                from vacancies 
                GROUP BY name
                ORDER BY COUNT(id) DESC
                LIMIT 10
            '''
            
df = pd.read_sql_query(query_7_1, connection)
df

C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\1199254702.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_1, connection)


name  count
0           Системный администратор   1761
1                    Программист 1С   1153
2                   Бизнес-аналитик    488
3  Специалист технической поддержки    457
4                Системный аналитик    432
5                          Аналитик    404
6               Инженер-программист    389
7                    Программист 1C    360
8              Frontend-разработчик    242
9     Инженер технической поддержки    230

In [381]:
# текст запроса
query_7_2 = f'''select e.name, COUNT(v.id) as "количество вакансий"
                from employers e
                JOIN vacancies v ON v.employer_id = e.id
                WHERE v.schedule LIKE 'Удаленная работа'
                GROUP BY e.id
                ORDER BY 2 DESC
                LIMIT 10
            '''
            
df = pd.read_sql_query(query_7_2, connection)
df

C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\2709213320.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_2, connection)


name  количество вакансий
0                    Яндекс                 1690
1                   DataArt                  165
2                  Тинькофф                   95
3        Nitka Technologies                   77
4                 ООО ЕФИН                    71
5                Ростелеком                   70
6  Иннотех, Группа компаний                   66
7           Bell Integrator                   57
8                  Homework                   54
9            ЦИФРОВОЙ ВЫБОР                   50

In [382]:
# текст запроса
query_7_3 = f'''select e.name, COUNT(v.id) as "количество вакансий"
                from employers e
                JOIN vacancies v ON v.employer_id = e.id
                WHERE (v.name iLIKE '%data scientist%'
                    OR v.name iLIKE '%data science%'
                    OR v.name iLIKE '%исследователь данных%'
                    OR v.name iLIKE '%machine learning%'
                    OR v.name iLIKE '%машинн%обучен%'
                    OR (v.name LIKE '%ML%' AND v.name NOT ILIKE '%html%'))
                GROUP BY e.id
                ORDER BY 2 DESC
                LIMIT 10
            '''
            
df = pd.read_sql_query(query_7_3, connection)
df

C:\Users\Лена\AppData\Local\Temp\ipykernel_12224\1876002039.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_3, connection)


name  количество вакансий
0                   СБЕР                   37
1        Bell Integrator                   25
2         Банк ВТБ (ПАО)                   18
3                     VK                   15
4  Positive Technologies                   11
5         EvenBet Gaming                    9
6                 Яндекс                    9
7                МегаФон                    8
8               Andersen                    7
9                    МТС                    6